### ISSUE TITLE
RGB OLED: Blocks Remove All Code

### ISSUE THREAD

**Issue by CarsonGracey** [parallaxinc/BlocklyProp#601]
Using Google Chrome

This issue has shown up today for me where placing any OLED block deletes all generated code until I delete the blocks and replace them. This has even happened on new projects.

### Blocks

![oled broken](https://cloud.githubusercontent.com/assets/20099916/18015363/41b0157e-6b7c-11e6-8e94-cb8c2adffce3.PNG)

### Code (Nothing shows up)

```C

```

And here is what happens when I disable these blocks. Keep in mind I've had this happen with every OLED block today, not just these two in particular.

### Blocks

![oled working](https://cloud.githubusercontent.com/assets/20099916/18015342/28829e0a-6b7c-11e6-93ff-ac85a968cf21.PNG)

### Code

```C
#include "simpletools.h"



int main()
{

    while (1)
    {}

}
```

This has occasionally fixed it - I delete/disable every OLED block then view the generated code. What happens is all other sensor/control/operator/all other blocks show up in the code as expected. Then if I replace/enable the blocks it will sometimes (rarely but it has happened) add the generated code from the OLED blocks in, if not it will remove all generated code for me again OR it will just continue showing all other sensor/control/operator/all other blocks generated code.

Note: The code won't compile either when the OLED blocks delete it all and if for some reason it does work, saving the project and refreshing the page creates this problem as well.

**Comment by kgracey** [comment]
@CarsonGracey asked me to verify the problem so I created this project http://demo.blockly.parallax.com/blockly/editor/blocklyc.jsp?project=301#

Initally, it showed the code. Then I saved it, went back to the blocks (which were all there) and the code was gone!

**Comment by PropGit** [comment]
@CarsonGracey - This is strange.  I haven't seen this exact behavior, but I have noticed that strange things happen when my testing overlaps system updates being applied to Demo.  Very frustrating.

I'm using Chrome, and yesterday, in nearly every strange case that I witnessed, simply pressing CTRL+refresh (to refresh without cache) solved it.  I'm sure you know about the need to do cachless refreshes, but I'm just reminding you in case you forgot this time.  Different browsers use different key combinations to do cachless refreshes; you'll have to make sure to use the right combination for yours.  If that doesn't solve it, then it's likely not a test-overlaps-system-update issue and it's likely a persistent problem.

**Comment by PropGit** [comment]
@kgracey - What browser are you using?

**Comment by kgracey** [comment]
Chrome Version 52.0.2743.116 (64-bit)

**Comment by CarsonGracey** [comment]
@PropGit, even after doing a cachless refresh I still ended up with the same block and code behavior.

**Comment by PropGit** [comment]
@kgracey @CarsonGracey - same version of Chrome I'm using.  This looks like a solid bug.

**Comment by kgracey** [comment]
I tried clearing cache and this problem persists.

**Comment by MatzElectronics** [comment]
I just tried too - nothing...

**Comment by zfi** [comment]
The thing that changed was the new Simple Library updates last night. I'm looking at this now.

**Comment by PropGit** [comment]
I think the issue they are experiencing is probably due to something that was changed earlier than the Simple Library updates, but not noticed until now.

**Comment by zfi** [comment]
I rolled back through a dozen commits and found that the issue appears at commit 25a00f99cf2c0cdfcdea10c7bd18aca7a934937d. This is when the new color picker was introduced. I am going to ask @DarkmatterVale to take a look at it.

**Comment by michel-cf** [comment]
![image](https://cloud.githubusercontent.com/assets/5745991/18018502/a5c89d96-6bd7-11e6-9d81-303b8f806a05.png)

oled.js:548 -> oled.js:548 Uncaught TypeError: Cannot read property '1' of null

**Comment by zfi** [comment]
What we seem to have is the Blockly.propc.valueToCode() call is no longer returning the expected value and the regex is no longer parsing the return value correctly.

Adding @DarkmatterVale to this to see if he can shed some light on what changed to cause this issue. Right now, our fallback position is to regress to the color picker from the Blockly base.

**Comment by zfi** [comment]
Update:

The Blockly.propc.valueToCode() appears to be returning the RGB color as a hexidecimal value and not a string of 6 characters. For example, it used to return 33CCFF and now returns 0x33CCFF.

I am refactoring the affected functions and should have a commit ready within the hour.

**Comment by zfi** [comment]
New code is in place on the Demo server.

**Comment by DarkmatterVale** [comment]
I'm sorry I was not online to fix this. Thank you @zfi for taking up the extra work!

**Comment by zfi** [comment]
Tracking this down forced me to actually use the embedded debugger in Firefox. Learned something new. 

@DarkmatterVale , you should take a look at the changes as it relates to what is getting returned from the color picker. I do not yet have an explanation as to why getFieldValue is returning a format of '#FFCC33' whereas calls to Blockly.propc.valueToCode() return a value in the format of '0xFFCC33'.  I suspect that the getFieldValue() calls are not using the same color picker code that the ...valueToCode() functions are. 

    // This returns a value in the form '#FF33CC', not '0xFF33CC'
     var color_mask = this.getFieldValue('flood');

    // This returns a value in the form '0xFF33CC'
    var color_mask = Blockly.propc.valueToCode(this, 'COLOR', Blockly.propc.ORDER_NONE);

**Comment by DarkmatterVale** [comment]
@zfi The reason getFieldValue returns a different output is because it does returns the selection of the color field, while valueToCode returns what I've done to the color (make it an integer). This creates a problem because we need to be able to store colors in variables; for that reason, we need to use valueToCode and not getFieldValue. Further, the ```flood``` input does not exist anymore; I replaced it with the value input ```COLOR``` (to more accurately reflect the input).

I will correct the OLED blocks later today and send a PR.

**Comment by DarkmatterVale** [comment]
The blocks should work fine after #609 is merged in

**Comment by DarkmatterVale** [comment]
@zfi @PropGit Did you update the SimpleLibraries on the demo server to include the new color math library?

**Comment by kgracey** [comment]
The problem with vaporizing code has been fixed. Several other problems exist which @DarkmatterVale can fix. I'll use some actual Blockly code to explain. 

![screen shot 2016-08-27 at 1 12 17 pm](https://cloud.githubusercontent.com/assets/9534148/18029820/eba090fa-6c57-11e6-9dbb-989e7c655bc7.png)

And the code generated by these blocks:

```
#include "simpletools.h"
#include "oledc.h"
#include "colormath.h"

int main()
    {
        oledc_init(4, 5, 6, 7, 8, 2);
        oledc_setCursor(0, 0, 0);
        oledc_drawText("Hello");
        oledc_drawLine(0, 0, 50, 50, oledc_color565(get8bitcolor(0xcc0000, "RED"), get8bitcolor(0xcc0000, "GREEN"), get8bitcolor(0xcc0000, "BLUE"));
        }
```

Has the following errors (which @DarkmatterVale and I verified in SimpleIDE):
1. Include file colormath.h should be colorMath.h (something needs to change - the generated code or the actual file name on the server)
2. A final ")" is needed on the long line of code starting with oledc_draw....
3. The get8bitcolor should be changed to get8bitColor

Then, everything compiles and produces the result I expected on the OLED. 

As we're getting into file naming conventions here I'm wondering if it's best to fix the file names for #1 and #3 above rather than the code the blocks create.

**Comment by zfi** [comment]
The latest libraries are on the Demo server. Last update was Thursday PM.

**Comment by kgracey** [comment]
OK - works. One thing we need to be aware of is that we are now using some camel-case naming conventions, so the blocks are getting adjusted to reflect this addition. If we don’t want that we better make a few changes (i.e. colormath.h is currently colorMath.h).

**Comment by MatzElectronics** [comment]
@zfi @PropGit It should be colormath (all lower)...I thought Andy and I had
that fixed, but apparently not...

It should be repaired in Simple Libraries if it's doing this...

**Comment by MatzElectronics** [comment]
Functions should be camel case, libraries should not...

**Comment by MatzElectronics** [comment]
In case anyone is up for an improvement, this:

`oledc_color565(get8bitcolor(0xcc0000, "RED"), get8bitcolor(0xcc0000, "GREEN"), get8bitcolor(0xcc0000, "BLUE"))`

could be changed to:

`remapColor(0xcc000, "8R8G8B", "5R6G5B");`

**Comment by DarkmatterVale** [comment]
@MatzElectronics changing it to remapColor() would cause more harm than good I think because the confusing use will not be easy for beginners to understand, while the get8bitColor() is much simpler.

Thoughts?

**Comment by MatzElectronics** [comment]
No preference, just a thought :)

**Comment by MatzElectronics** [comment]
@kgracey If it's not too much trouble, can you grab a new copy of Simple Libraries so that you are using the exact copy that Blockly is?  This should correct the "colormath" problem...

It's here:
https://github.com/parallaxinc/Simple-Libraries/archive/master.zip

**Comment by MatzElectronics** [comment]
@PropGit @zfi @DarkmatterVale @kgracey  Update - the colormath library Andy and I put up into SimpleLibraries, that was then used in Blockly is corrupt.  Took me a bit to pin it down, but I fixed it and and emailed a new one to Jeff and Jim to post to GitHub.  I also submitted the PR to revert Vale's changes back to "colormath.h" as they should properly be.

I apologize for sending up a broken library in the first place. Thanks,
Matt
